File: Graph/BipartiteGraph/bip.py
from collections import deque

class Solution:
    def colorByBFS(self, adjList, start, color):
        color[start] = 0
        qu = deque()
        qu.append(start)
        while len(qu):
            curr = qu.popleft()
            for neighbour in adjList[curr]:
                if color[neighbour] == -1:
                    color[neighbour] = 1-color[curr]
                    qu.append(neighbour)
                elif color[neighbour] == color[curr]: return False
        return True
        
    def isBipartite(self, V, edges):
        # code here
        color = [-1] * V
        adjList = [[] for _ in range(V)]
        totalEdges = len(edges)
        for i in range(totalEdges):
            u = edges[i][0]
            v = edges[i][1]
            adjList[u].append(v)
            adjList[v].append(u)
        for i in range(V):
            if color[i]==-1:
                if not self.colorByBFS(adjList,i,color): return False
                
        return True

File: Graph/BipartiteGraph/bip.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    bool colorByBFS(vector<vector<int>>& adjList, int start, vector<int>& color) {
        color[start] = 0;

        queue<int> qu;
        qu.push(start);

        while (!qu.empty()) {
            int curr = qu.front();
            qu.pop();

            for (int neighbour : adjList[curr]) {
                if (color[neighbour] == -1) {
                    color[neighbour] = 1 - color[curr];
                    qu.push(neighbour);
                }
                else if (color[neighbour] == color[curr]) {
                    return false;
                }
            }
        }

        return true;
    }

    bool isBipartite(int V, vector<vector<int>>& edges) {
        vector<int> color(V, -1);
        vector<vector<int>> adjList(V);

        for (auto &edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adjList[u].push_back(v);
            adjList[v].push_back(u);
        }

        for (int i = 0; i < V; i++) {
            if (color[i] == -1) {
                if (!colorByBFS(adjList, i, color))
                    return false;
            }
        }

        return true;
    }
};